File: backend/src/main/resources/static/app/still-true.js
/* "Still true?" — the records nobody has confirmed in a while (docs/21).

   A card on Home rather than a screen of its own: the question is small, and a
   yes should take one tap from the place people already look. It shows only
   what the server says this person is being asked about, which is only what
   they own or hold; nobody is asked to vouch for someone else's policy.

   The server answers an older client's missing endpoint with a 404, and Home
   must not break for that, so a failed load is simply no card. */

import { api } from "./api.js";
import { el, toast } from "./ui.js";
import { t, localDate } from "./i18n.js";

/** The list, or nothing. Never throws: this card is never why Home fails. */
export async function loadStillTrue(householdId) {
  try {
    const result = await api.stillTrue(householdId);
    return Array.isArray(result?.items) ? result.items : [];
  } catch {
    return [];
  }
}

/** One month on, as the household's calendar date the server validates. */
export function monthFromToday(now = new Date()) {
  const next = new Date(now.getFullYear(), now.getMonth() + 1, now.getDate());
  // 31 January plus a month is not 3 March: clamp to the month's last day.
  if (next.getDate() !== now.getDate()) next.setDate(0);
  const pad = (n) => String(n).padStart(2, "0");
  return `${next.getFullYear()}-${pad(next.getMonth() + 1)}-${pad(next.getDate())}`;
}

export function stillTrueCard(householdId, items) {
  if (!items.length) return null;

  const list = el("div.list", {});
  const count = el("span.caption.muted", {}, String(items.length));
  const card = el("div.card", {},
    el("div.section-title", {}, el("h4", {}, t("still.title")), count),
    el("p.caption.muted", { style: { marginTop: 0 } }, t("still.intro")),
    list,
  );

  const remove = (row) => {
    row.remove();
    count.textContent = String(list.children.length);
    if (!list.children.length) card.remove();
  };

  items.forEach((item) => {
    const buttons = [];
    const busy = (on) => buttons.forEach((b) => { b.disabled = on; });

    const confirm = el("button.btn.btn-sm.btn-primary", {
      type: "button",
      onclick: async () => {
        busy(true);
        try {
          const answered = await api.confirmStillTrue(householdId, item.recordType, item.recordId);
          toast(t("still.confirmed", { date: localDate(answered.dueOn) }));
          remove(row);
        } catch (error) {
          busy(false);
          toast(error.message, { tone: "error" });
        }
      },
    }, t("still.confirm"));

    const later = el("button.btn.btn-sm", {
      type: "button",
      onclick: async () => {
        busy(true);
        try {
          const snoozed = await api.snoozeStillTrue(
            householdId, item.recordType, item.recordId, monthFromToday());
          toast(t("still.snoozed", { date: localDate(snoozed.dueOn) }));
          remove(row);
        } catch (error) {
          busy(false);
          toast(error.message, { tone: "error" });
        }
      },
    }, t("still.snooze"));
    buttons.push(confirm, later);

    const why = item.reason === "key_date" && item.keyDate
      ? t("still.keyDate", { date: localDate(item.keyDate) })
      : t("still.lastConfirmed", { date: localDate(item.lastConfirmedAt) });

    // The answers sit under the title rather than beside it: beside it, on a
    // phone, two buttons squeezed the title to one word a line.
    const row = el("div.list-row", { style: { alignItems: "flex-start" } },
      el("span.pill", {}, t(`where.type.${item.recordType}`)),
      el("div.grow", {},
        el("div.title", {}, item.title),
        el("div.meta", {}, why),
        el("div.row.wrap", { style: { gap: "8px", marginTop: "10px" } }, confirm, later),
      ),
    );
    list.append(row);
  });

  return card;
}

File: backend/src/main/kotlin/tech/bhrigu/almira/stilltrue/StillTrue.kt
package tech.bhrigu.almira.stilltrue

import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import tech.bhrigu.almira.audit.AuditService
import tech.bhrigu.almira.common.ApiException
import tech.bhrigu.almira.household.HouseholdService
import tech.bhrigu.almira.security.RequestUserContext
import java.time.Instant
import java.time.LocalDate
import java.util.UUID

/**
 * One record the caller is being asked about (docs/21).
 *
 * [reason] says why now: `period` when it has simply not been confirmed for
 * [periodMonths], `key_date` when a maturity, renewal or end date has passed
 * since the last confirmation — the moment the answer may have changed.
 */
data class StillTrueItem(
    val recordType: String,
    val recordId: UUID,
    /** Not sealed: the caller can see the record, and titles drive every list. */
    val title: String,
    val periodMonths: Int,
    /** The later of creation, the last value refresh and the last confirmation. */
    val lastConfirmedAt: Instant,
    val keyDate: LocalDate?,
    val dueOn: LocalDate,
    val snoozedUntil: LocalDate?,
    val isDue: Boolean,
    val reason: String,
)

data class StillTrueList(val items: List<StillTrueItem>)

/** Named apart from the reminder's snooze body: OpenAPI schema names are global. */
data class StillTrueSnoozeBody(val until: LocalDate)

object StillTrue {
    val RECORD_TYPES = listOf("investment", "liability", "account", "estate_document")

    /** The furthest a question can be put off. Beyond a year it is not a snooze. */
    const val MAX_SNOOZE_DAYS = 366L

    /** What a notification says. A count, never a title or an amount. */
    fun digestTitle(count: Int): String =
        if (count == 1) "Still true? 1 record to check" else "Still true? $count records to check"

    const val DIGEST_TEMPLATE = "still_true.digest"
}

/**
 * "Still true?" for a person, through their own row-level security.
 *
 * Everything is read from `still_true_items`, the same view the sweep reads, so
 * what the list shows and what a notification counted cannot drift apart.
 */
@Service
class StillTrueService(
    private val jdbc: NamedParameterJdbcTemplate,
    private val households: HouseholdService,
    private val audit: AuditService,
    private val userContext: RequestUserContext,
) {

    @Transactional(readOnly = true)
    fun due(householdId: UUID): StillTrueList {
        userContext.require()
        households.get(householdId)
        val items = jdbc.query(
            "$SELECT where household_id = :hid and is_due order by effective_due_on, lower(title), record_id",
            mapOf("hid" to householdId),
        ) { rs, _ -> item(rs) }
        return StillTrueList(items)
    }

    /**
     * "Yes, this is still true."
     *
     * Only the people asked may answer — the owners, holders or the person whose
     * will it is. Anyone else gets the same 404 as for a record they cannot see:
     * being able to read a household-visible record is not knowing whether it is
     * still in force.
     */
    @Transactional
    fun confirm(householdId: UUID, recordType: String, recordId: UUID): StillTrueItem {
        val userId = requireWriter()
        households.get(householdId)
        find(householdId, recordType, recordId)

        jdbc.update(
            """
            insert into record_confirmations (record_type, record_id, household_id, confirmed_at, confirmed_by)
            values (:type, :id, :hid, now(), :uid)
            on conflict (record_type, record_id) do update
              set confirmed_at = now(), confirmed_by = :uid,
                  -- An answer ends a snooze: there is nothing left to put off.
                  snoozed_until = null, snoozed_by = null, updated_at = now()
            """.trimIndent(),
            params(householdId, recordType, recordId).addValue("uid", userId),
        )
        // The holding's own "Last confirmed" line and the dashboard's freshness
        // card read this column. One answer should move all three.
        if (recordType == "investment") {
            jdbc.update(
                "update investments set last_verified_at = now() where id = :id and deleted_at is null",
                mapOf("id" to recordId),
            )
        }
        audit.record(
            householdId = householdId, actorUserId = userId, action = "record.confirm_still_true",
            entityType = recordType, entityId = recordId,
        )
        return find(householdId, recordType, recordId)
    }

    /** "Ask me later." Moves the question; says nothing about the answer. */
    @Transactional
    fun snooze(householdId: UUID, recordType: String, recordId: UUID, until: LocalDate): StillTrueItem {
        val userId = requireWriter()
        households.get(householdId)
        val current = find(householdId, recordType, recordId)
        // "Today" where the family lives, which is what due dates are in.
        val today = jdbc.queryForObject(
            "select (now() at time zone time_zone)::date from households where id = :hid",
            mapOf("hid" to householdId), LocalDate::class.java,
        )!!
        if (!until.isAfter(today)) {
            throw ApiException.badRequest("snooze_past", "Choose a date after today.")
        }
        if (until.isAfter(today.plusDays(StillTrue.MAX_SNOOZE_DAYS))) {
            throw ApiException.badRequest(
                "snooze_too_far", "Choose a date within a year. Longer than that isn't a snooze.",
            )
        }
        jdbc.update(
            """
            insert into record_confirmations (record_type, record_id, household_id, snoozed_until, snoozed_by)
            values (:type, :id, :hid, :until, :uid)
            on conflict (record_type, record_id) do update
              set snoozed_until = :until, snoozed_by = :uid, updated_at = now()
            """.trimIndent(),
            params(householdId, recordType, recordId).addValue("until", until).addValue("uid", userId),
        )
        audit.record(
            householdId = householdId, actorUserId = userId, action = "record.snooze_still_true",
            entityType = recordType, entityId = recordId,
            diff = mapOf("from" to current.snoozedUntil?.toString(), "to" to until.toString()),
        )
        return find(householdId, recordType, recordId)
    }

    private fun requireWriter(): UUID {
        val userId = userContext.require()
        // A guest link is read-only in the database's own terms already; saying
        // so here turns a failed statement into a sentence.
        if (userContext.currentGuestShareId() != null) throw ApiException.forbidden()
        return userId
    }

    private fun find(householdId: UUID, recordType: String, recordId: UUID): StillTrueItem {
        if (recordType !in StillTrue.RECORD_TYPES) {
            throw ApiException.badRequest(
                "record_type_invalid", "Choose one of: ${StillTrue.RECORD_TYPES.joinToString()}.",
            )
        }
        return jdbc.query(
            "$SELECT where household_id = :hid and record_type = :type and record_id = :id",
            params(householdId, recordType, recordId),
        ) { rs, _ -> item(rs) }.firstOrNull() ?: throw ApiException.notFound()
    }

    private fun params(householdId: UUID, recordType: String, recordId: UUID) =
        MapSqlParameterSource().addValue("hid", householdId)
            .addValue("type", recordType).addValue("id", recordId)

    private fun item(rs: java.sql.ResultSet): StillTrueItem {
        val keyDate = rs.getObject("key_date", LocalDate::class.java)
        val dueOn = rs.getObject("due_on", LocalDate::class.java)
        return StillTrueItem(
            recordType = rs.getString("record_type"),
            recordId = rs.getObject("record_id", UUID::class.java),
            title = rs.getString("title"),
            periodMonths = rs.getInt("period_months"),
            lastConfirmedAt = rs.getTimestamp("last_confirmed_at").toInstant(),
            keyDate = keyDate,
            dueOn = rs.getObject("effective_due_on", LocalDate::class.java),
            snoozedUntil = rs.getObject("snoozed_until", LocalDate::class.java),
            isDue = rs.getBoolean("is_due"),
            reason = if (keyDate != null && dueOn == keyDate.plusDays(KEY_DATE_GRACE_DAYS)) "key_date" else "period",
        )
    }

    private companion object {
        const val SELECT = """
            select record_type, record_id, title, period_months, last_confirmed_at, key_date,
                   due_on, snoozed_until, effective_due_on, is_due
            from still_true_items
        """

        /** Mirrors `key_date + 7` in V29's still_true_records. */
        const val KEY_DATE_GRACE_DAYS = 7L
    }
}

@RestController
@RequestMapping("/api/v1/households/{householdId}/still-true")
class StillTrueController(private val service: StillTrueService) {

    /** What the caller is being asked about now: due, and theirs to answer. */
    @GetMapping
    fun due(@PathVariable householdId: UUID): StillTrueList = service.due(householdId)

    @PostMapping("/{recordType}/{recordId}/confirm")
    fun confirm(
        @PathVariable householdId: UUID,
        @PathVariable recordType: String,
        @PathVariable recordId: UUID,
    ): StillTrueItem = service.confirm(householdId, recordType, recordId)

    @PostMapping("/{recordType}/{recordId}/snooze")
    fun snooze(
        @PathVariable householdId: UUID,
        @PathVariable recordType: String,
        @PathVariable recordId: UUID,
        @RequestBody body: StillTrueSnoozeBody,
    ): StillTrueItem = service.snooze(householdId, recordType, recordId, body.until)
}
